server/lib/src/idm/oauth2.rs

```diff
@@ -926,7 +979,119 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
                 refresh_token,
                 scope,
             } => self.check_oauth2_token_refresh(o2rs, refresh_token, scope.as_ref(), ct),
+            GrantTypeReq::DeviceCode { device_code, scope } => {
+                self.check_oauth2_device_code_status(device_code, scope)
+            }
+        }
+    }
+
+    #[instrument(level = "info", skip(self))]
+    pub fn handle_oauth2_start_device_flow(
+        &mut self,
+        client_auth_info: ClientAuthInfo,
+        client_id: &str,
+        scope: &Option<BTreeSet<String>>,
+        eventid: Uuid,
+    ) -> Result<DeviceAuthorizationResponse, Oauth2Error> {
+        // DANGER: Why do we have to do this? During the use of qs for internal search
+        // and other operations we need qs to be mut. But when we borrow oauth2rs here we
+        // cause multiple borrows to occur on struct members that freaks rust out. This *IS*
+        // safe however because no element of the search or write process calls the oauth2rs
+        // excepting for this idm layer within a single thread, meaning that stripping the
+        // lifetime here is safe since we are the sole accessor.
+        let o2rs: &Oauth2RS = unsafe {
+            let s = self.oauth2rs.inner.rs_set.get(client_id).ok_or_else(|| {
+                admin_warn!("Invalid OAuth2 client_id {}", client_id);
+                Oauth2Error::AuthenticationRequired
+            })?;
+            &*(s as *const _)
+        };
+
+        info!("Got Client: {:?}", o2rs);
+
+        if !o2rs.require_pkce() {
+            security_info!("Device flow is only available for PKCE-enabled clients");
+            return Err(Oauth2Error::InvalidRequest);
         }
+
+        info!(
+            "Starting device flow for client_id={} scopes={} source={:?}",
+            client_id,
+            scope
+                .as_ref()
+                .map(|s| s.iter().cloned().collect::<Vec<_>>().into_iter().join(","))
+                .unwrap_or("[]".to_string()),
+            client_auth_info.source
+        );
+
+        let mut verification_uri = self.oauth2rs.inner.origin.clone();
+        verification_uri.set_path(uri::OAUTH2_DEVICE_LOGIN);
+
+        // device code is a combination of the inputs
+        fn gen_device_code(user_code: &str, expiry: &Duration) -> Result<String, Oauth2Error> {
+            let mut hasher = Sha256::new();
+            hasher.update(user_code.as_bytes());
+            hasher.update(expiry.as_secs().to_be_bytes());
+            let hash = hasher.finalize();
+            Ok(general_purpose::STANDARD_NO_PAD.encode(hash))
+        }
+
+        fn gen_user_code() -> Result<String, Oauth2Error> {
```

Comment:
I think we should limit to numeric only, and we should hyphenate on triples. This makes it easier to read the code and enter it on a pin pad, phone or similar. Per the rfc this gives 012-324-567 as the code style, and it has 10^9 - 1 bits of entropy, meaning that an attacker would need to sample and attempt 3 million codes per second to bruteforce the whole space. In the db I think we should store the code as a uint32 since we are doing numeric only.